### DISCUSSION TITLE
List of djot implementations

### DISCUSSION THREAD

**Discussion by jgm** [jgm/djot#265] [General]
I'd like to make a list of working djot implementations.

I'm responsible for three:

- https://github.com/jgm/djot.lua (Lua)
- https://github.com/jgm/djot.js (Typescript/Javascript)
- https://github.com/jgm/djoths (Haskell, creates AST)

I also know of:

- https://github.com/aarroyoc/djota (Prolog, by @aarroyoc)
- https://github.com/hellux/jotdown (Rust, pull parser, by @hellux)

If you've got an implementation that isn't on this list, post it here!

**Comment by faelys** [comment]
I have https://github.com/faelys/djot.erlang which is almost finished but still fails a few corner cases, and the code is very ugly. I'm planning to finish and refactor it in a few weeks or so.

I also started a parser in chicken scheme, based on a PEG combinator library, but it's not far enough to be sure it's actually possible to parse proper djot with this framework, so I haven't published yet. I'll let you know if it ever reaches a usable state.

**Comment by sivukhin** [comment]
I have implemented parser in Go: https://github.com/sivukhin/godjot

**Comment by paradox460** [comment]
I've built [djot for elixir](https://github.com/paradox460/djot), which is just a NIF wrapper around the rust jotdown crate

**Comment by dereuromark** [comment]
PHP: https://github.com/php-collective/djot-php

Docs via vitepress auto-generated: https://php-collective.github.io/djot-php/

**Comment by dereuromark** [comment]
Small benchmark of the different languages (or implementations maybe):


| Implementation      | Mean Time | Throughput | vs PHP    |
|---------------------|-----------|------------|-----------|
| Rust (jotdown)      | ~1-2 ms   | ~30+ MB/s  | ~10x faster |
| Go (godjot)         | ~2-4 ms   | ~15+ MB/s  | ~5x faster |
| JS (@djot/djot)     | 8.1 ms    | 5.2 MB/s   | 2.2x faster |
| PHP (djot-php)      | 18.1 ms   | 3.0 MB/s   | baseline  |
| Python-Markdown*    | 41.1 ms   | 1.0 MB/s   | 2.3x slower |
| markdown-it-py*     | 36.8 ms   | 1.2 MB/s   | 2.0x slower |

https://github.com/php-collective/djot-php/blob/master/docs/performance.md#cross-language-comparison

**Comment by dereuromark** [comment]
Another interest benchmark is between PHP parsers (same language):


With equivalent features enabled (tables, footnotes, smart typography):

| Library | 30KB Doc | Throughput | vs djot-php |
|---------|----------|------------|-------------|
| erusev/parsedown | 1.69 ms | 16.0 MB/s | 6.7x faster |
| michelf/php-markdown | 5.16 ms | 5.2 MB/s | 2.2x faster |
| michelf/php-markdown (Extra) | 6.15 ms | 4.4 MB/s | 1.9x faster |
| **djot-php** | **11.36 ms** | **2.6 MB/s** | baseline |
| league/commonmark (GFM) | 15.00 ms | 1.8 MB/s | 1.3x slower |
| league/commonmark | 15.31 ms | 1.8 MB/s | 1.4x slower |
| league/commonmark (Full) | 23.54 ms | 1.3 MB/s | **2.1x slower** |

**Key finding:** djot-php is **2x faster than CommonMark** when both have equivalent features (tables, footnotes, smart punct) enabled.

It shows that Djot sure is designed to parse more efficient.

https://github.com/php-collective/djot-php/blob/master/docs/performance.md#php-alternatives-comparison

**Comment by dereuromark** [comment]
:tada:  Djot is now officially available worldwide as plugin in Wordpress installs: https://wordpress.org/plugins/djot-markup/
Please feel free to test it, review and rate it so people have confidence in switching over from old markdown or other systems.

Most notably,  it now contains also useful addons like
- TOC (auto generated)
- Permalinks (clickable/copyable via hover icon)
- Locale-aware smart quotes
- Enhanced syntax (e.g. Markdown compatibility mode for easier upgrading/switching)
- Permissions per spec element (author vs commenter etc)

and more.

**Comment by karlb** [comment]
I've create a C djot->HTML converter: https://github.com/karlb/cdjot

Except for the djot.js test suite, is there anything I should test this against? Do we have a benchmark for djot implementations?

**Reply by dereuromark** [reply]
I recommend testing against full specs.
You can use the official djot specs of the djot.js repo, and on top I created https://github.com/php-collective/djot-php/tree/master/tests/official that has the same bases with a bunch of more incl edge cases.

**Reply by hellux** [reply]
There are some benchmarks at <https://github.com/dcampbell24/djot-implementations>

**Comment by dereuromark** [comment]
https://github.com/dereuromark/awesome-djot
This list is useful as up to date overview and could be pulled into an org if there was one to be created for djot.
